### ISSUE TITLE
Non-English posts slip through: language filtering trusts absent tags as English

### ISSUE THREAD

**Issue by bmilojkovic** [goodgorithm/goodgorithm#28]
## Scope

The feed is meant to be English-only (`CLAUDE.md`: "English-only on both paths; every downstream model is English-only"), but real production sampling shows a confirmed, non-trivial share of non-English posts making it through on both sources.

## The gap, confirmed against production

`ingestion/src/bluesky.ts`: `if (langs.length > 0 && !langs.includes("en")) return;` — only excludes when the post *has* a self-reported `langs` array that doesn't include `en`. An **empty** `langs` array (no self-tag at all) bypasses the check entirely and is kept unconditionally.

`ingestion/src/mastodon.ts`: `if (status.language && status.language !== "en") continue;` — same shape of gap. A **null** `status.language` (Mastodon's own server-side detection didn't populate it) is kept unconditionally.

Both paths treat "no language signal" as "assume English" rather than "verify or exclude." Queried `raw_posts` directly:

```
source    null_lang  total  pct_null
mastodon  5,477      49,695 11.0%
bluesky   6,957      55,389 12.6%
```

~11-13% of all ingested posts on both sources carry no language tag at all. A random sample of 25 `lang IS NULL` rows found roughly 40%+ were clearly **not** English — Japanese and Spanish were the most common (e.g. "もう今日なんも食えねえや...", "10年前くらいから成ってたんすよね...", "Vivir en flores, camino ajeno...", "Espectacular eclipse total de Sol en Ibiza..."), alongside genuinely-English untagged posts. So this isn't a rare edge case — it's a real, meaningfully-sized slice of total ingested volume (rough order of magnitude: a few percent of everything ingested), and everything that gets through it flows straight into a pipeline (topicality, sentiment, taxonomy) that's English-only by design and presumably degrades silently on other languages rather than erroring.

(Smaller, secondary wrinkle also visible in the data: a handful of posts stored with `lang` like `es`/`de`/`pt`/`fr` despite the `en`-inclusion check passing — likely multi-tagged/bilingual posts where `en` is present in `langs` alongside another language, and `lang: langs[0] ?? null` just happened to store the non-`en` one first. Probably fine as-is — a genuinely bilingual post including English isn't the same problem — but worth a quick look alongside the main fix.)

## Suggested direction (research done, not a prescribed fix)

The absent-tag case needs actual content-based language verification, not just trusting metadata — classical, non-LLM, consistent with `CLAUDE.md`'s no-LLM-in-the-algorithm constraint (this is a hard eligibility filter, the same category as `content_filter.py`/`bot_filter.py`, not scoring/ranking). Two well-established, lightweight, locally-run options worth benchmarking against a real sample of this project's own untagged posts before picking one:

- **fastText's `lid.176`** — Meta's pretrained language-ID model, 176 languages, trained on Wikipedia/Tatoeba/SETimes. Ships in two sizes: `lid.176.bin` (126MB, marginally more accurate) or `lid.176.ftz` (917KB compressed, small accuracy tradeoff) — the compressed version is small enough to not need `processing/`'s R2 model-hosting pattern at all, just a bundled dependency. Widely used for exactly this (tweet/short-text language ID at scale), but general accuracy claims are for longer text — worth specifically checking behavior on short, hashtag/emoji-heavy posts like this project actually ingests.
- **`lingua-py`** — ~75 languages, positioned specifically for short text (titles, queries, short messages) rather than document-length text, which matches this project's actual input shape (posts, not articles) more closely than fastText's typical benchmark conditions. Worth a head-to-head comparison against fastText on a real sample rather than assuming either wins.

A cheap Unicode-script pre-check (e.g. non-Latin-script character ratio) could catch some non-English content (Japanese, Cyrillic, Arabic, etc.) at near-zero cost before ever reaching a model, but wouldn't catch same-script false negatives — the Spanish/French/German/Portuguese posts in the sample above are all Latin-script and indistinguishable from English by script alone, so a real classifier is still needed regardless.

**Where this runs**: `processing/` (Python) is the natural fit over `ingestion/` (TypeScript) — it's where all the other classical-ML filtering already lives (spaCy, sklearn, VADER/CNN), and `pipeline.py`'s `run_cycle` already has a "hard exclude before dedup/bot/topicality ever run" gate (`content_filter.is_content_excluded()`, and now also issue #7's blocklist check) that a language-verification step could slot into directly — cheapest place to add a new exclude, same established pattern. Only running it on the ~11-13% of posts with no/ambiguous tag (rather than all posts) keeps the added cost bounded to roughly the size of the confirmed gap.

## Where to look

- `ingestion/src/bluesky.ts`, `ingestion/src/mastodon.ts` — the current tag-trusting filters.
- `processing/src/content_filter.py`, `processing/src/pipeline.py`'s `run_cycle` — the existing hard-exclude gate pattern (and issue #7's blocklist check, same position) this would extend.
- `processing/src/model_store.py` — the existing R2-hosted-model pattern, if the larger `lid.176.bin` variant or a lingua-py model ends up needing it (likely unnecessary given the `.ftz` variant's size).
- `CLAUDE.md`'s Data sources section (English-only claim) and `CONTENT_POLICY.md` — update once this ships, same as issue #7's instruction to update its own section.
